### DISCUSSION TITLE
Using Shoelace form components with Unpoly

### DISCUSSION THREAD

**Discussion by msolli** [unpoly/unpoly#643] [Q&A] [answered]
Hi all! I'm just getting started with Unpoly, and I'm having a great time with it, so thanks!

However I've hit a snag where I'm unsure how to troubleshoot:

I'm using Shoelace web components. In an Unpoly modal I have a form with Shoelace form components such as `<sl-input>` and `<sl-button>`. When the form is submitted using the **Enter** key, and there is server validation such that the server responds with 422 and a HTML response, the following error is printed in the console:
`Uncaught (in promise) up.CannotMatch: Could not find a layer to render in. You may have passed an unmatchable layer reference, or a detached element.`
When the form is submitted using a click on the `<sl-button>`, everything works as expected.

I have reproduced the problem in a Glitch: https://glitch.com/edit/#!/teal-spurious-join

To reproduce:
1. Open modal
2. Type anything but the letter 'x' in the form input under "Shoelace form compoents".
3. Submit form using the **Enter** key

What I expect:
That the modal updates with the HTML returned from the server. The message "Not x" should be displayed just under the text input.

What actually happens:
The form is submitted, but the response is not rendered. Instead there is an error in the browser console: `Uncaught (in promise) up.CannotMatch: Could not find a layer to render in. You may have passed an unmatchable layer reference, or a detached element.`

When the form is submitted using a click on the Submit button, I get the expected result.

I'd love to hear your thoughts on how I might approach this problem, and if any of you have gotten Shoelace form components to work with Unpoly?

**Comment by triskweline** [comment]
Your code looks good to me.

In my test the modal behaved as you expected:

![Screenshot from 2024-06-13 09-05-48](https://github.com/unpoly/unpoly/assets/89953/13e8390a-a472-4e52-ad68-f6aabf5d33d4)

> I'd love to hear your thoughts on how I might approach this problem

Try using `up.log.enable()` to enable verbose debugging output on the console.

**Comment by msolli** [comment]
Hi Henning, thanks for looking into this! I noticed now that my description was incomplete - I had written `<Enter>`, which was converted to an HTML tag by the Markdown parser, and thus invisible to the reader. I've corrected the text now.

The crucial point is that the problem appears only when the form is submitted using the Enter key. If it submitted by clicking on the submit button it works fine.

Here is a screen recording that demonstrates it:

https://github.com/unpoly/unpoly/assets/70058/d3db2add-97f8-4861-af53-244185d052fa

(Tested in both Firefox and Chrome.)

**Reply by triskweline** [reply, marked as answer]
I can reproduce it, but not entirely sure what's going on there. I'm not very familiar with Shoelace's internals.

My guess is:

- When a form is submitted, Unpoly [remembers which element caused the submission](https://github.com/unpoly/unpoly/blob/788a6ddda1ddc4658d235cbfdffaa2be1aed47fc/src/unpoly/form.js#L366) as `{ origin }`. By default this is the focused element.
- The origin element is later used to decide which layer to render. Since no `[up-fail-layer]` is given, Unpoly will render the failed response within the origin's layer.
- Somehow the focused `<sl-input>` appears as detached for Unpoly. Maybe this is a shadow DOM thing, or maybe Shoelace swaps its `<sl-input>` for a new copy.

I was able to work around this by explicitly adding an `[up-fail-layer=current]` attribute to the overlay form. More analysis is welcome!

**Reply by msolli** [reply]
Thanks! I can confirm `[up-fail-layer=current]` works as a workaround, both in the example and and in my production code.

I'll try to run this by the Shoelace folks and see if they can make some sense of it.

**Reply by triskweline** [reply]
It [looks like](https://github.com/shoelace-style/shoelace/blob/96edd854f3cf8a79da5eff3f86f1489a7e3e9f75/src/internal/form.ts#L364-L366) Shoelace submits a form by: 

- creating a temporary submit button
- appending it to the form
- clicking it
- immediately removing it

That's why Unpoly sees a detached `{ origin }` one microtask later.

**Reply by msolli** [reply]
Thanks for looking into it! Would you still say that `[up-fail-layer]` is the preferred workaround, or are there other options?

**Reply by triskweline** [reply]
It is the preferred workaround for now. It should no longer be required with the next Unpoly release.

**Comment by triskweline** [comment]
This issue has been resolved with version 3.9.0.

**Reply by msolli** [reply]
Thanks! Confirming it works in the [Glitch](https://glitch.com/edit/#!/teal-spurious-join) and in my production code.

However, there is now a console error after the server fragment has rendered:
```
Uncaught (in promise) TypeError: this.input is null
    focus https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.15.1/cdn/chunks/chunk.ODJH6G2R.js:180
    doFocus https://cdn.jsdelivr.net/npm/unpoly@3.9.0/unpoly.js:8810
    restore https://cdn.jsdelivr.net/npm/unpoly@3.9.0/unpoly.js:3940
    _restoreLostFocus https://cdn.jsdelivr.net/npm/unpoly@3.9.0/unpoly.js:4126
    processPrimitive https://cdn.jsdelivr.net/npm/unpoly@3.9.0/unpoly.js:4089
...
```

It can be reproduced with original procedure in this discussion.

Is this yet another quirk of how Shoelaces does form submissions?

**Reply by triskweline** [reply]
:see_no_evil: 

Apparently Shoelace components sometimes crash when Unpoly calls `focus()` on them. Here: To restore focus after a fragment was swapped. Maybe Shoelace takes another task to initialize its elements.

Unfortunately I don't have the bandwidth to look into this right now.

**Reply by triskweline** [reply]
Analysis:

- When swapping out a focused form field with new content, Unpoly tries to rediscover the form field in the new content. If a similiar field has been found (e.g. a field with the same `[id]`) it refocuses the new field with `field.focus()`.
- When calling `shoelaceComponent.focus()` it delegates to `shoelaceComponent.input.focus()`.
- `shoelaceComponent.input` is `null` until the component has called `render()` internally
- Shoelace does not render synchronously after being inserted. Instead if waits for a microtask (maybe an `await` somewhere).
- Hence Unpoly calls `shoelaceComponent.focus()` before `shoelaceComponent.input` is defined, causing the error trace above.

The [test case](https://glitch.com/edit/#!/curvy-torch-slice) for this does the following:

```js
window.addEventListener('load', function() {
  let shoelaceInput = document.createElement('sl-input')
  document.body.append(shoelaceInput)
  
  console.log("[sync after insert] shoelaceInput.input is %o", shoelaceInput.input)

  Promise.resolve().then(function() {
    console.log("[next microtask] shoelaceInput.input is %o", shoelaceInput.input)
  })
  
  setTimeout(function() {
    console.log("[next task] shoelaceInput.input is %o", shoelaceInput.input)
  })
})
```

This logs:

```
[sync after insert] shoelaceInput.input is null
[next microtask] shoelaceInput.input is <input part=​"input" id=​"input" class=​"input__control" aria-describedby=​"help-text" type=​"text" title name placeholder spellcheck=​"true">​
[next task] shoelaceInput.input is <input part=​"input" id=​"input" class=​"input__control" aria-describedby=​"help-text" type=​"text" title name placeholder spellcheck=​"true">​
```

**Reply by triskweline** [reply]
Custom elements run `connectedCallback()` synchronously after insertion.

It seems to be a behavior of Lit (used by Shoelace) that `render()` is not called synchronously after insertion.

**Reply by triskweline** [reply]
[Lit Reactive update cycle](https://lit.dev/docs/components/lifecycle/#reactive-update-cycle):

> The reactive update cycle is triggered when a reactive property changes or when the requestUpdate() method is explicitly called. Lit performs updates asynchronously so property changes are batched — if more properties change after an update is requested, but before the update starts, all of the changes are captured in the same update.
> 
> Updates happen at microtask timing, which means they occur before the browser paints the next frame to the screen. See [Jake Archibald's article](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) on microtasks for more information about browser timing.

**Reply by triskweline** [reply]
I opened a Shoelace issue here: https://github.com/shoelace-style/shoelace/issues/2176

**Reply by msolli** [reply]
Thanks again for looking into this! If I understand correctly, this is tied to how Custom Elements work, maybe, or how Lit elements work. Given that other libraries and approaches for creating Custom Elements might have still other ways of initializing and managing lifecycle, maybe Unpoly could provide an extension point for initializing Custom Elements? Something akin to `config.fieldSelectors`, where one could provide callbacks to be called when a Custom Element (like `sl-input`) is inserted into the DOM.

**Reply by triskweline** [reply]
> maybe Unpoly could provide an extension point for initializing Custom Elements?

Maybe, yes. We would need to go through all the code where Unpoly interfaces with custom elements (e.g. focusing, watching) and wait for the hook (or `customElements.whenDefined()`). The painful part is that this would make a lot of operations async which are now sync, allowing concurrency issues into what is now simple code.
